File: code/packages/python/compute-runtime/tests/test_memory.py
"""Tests for MemoryManager, Buffer, MappedMemory."""

import pytest

from compute_runtime import (
    RuntimeInstance,
    MemoryType,
    BufferUsage,
    Buffer,
    MappedMemory,
)


def make_device(vendor: str = "nvidia"):
    instance = RuntimeInstance()
    physical = next(
        d for d in instance.enumerate_physical_devices()
        if d.vendor == vendor
    )
    return instance.create_logical_device(physical)


class TestAllocate:
    def test_basic_allocation(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(256, MemoryType.DEVICE_LOCAL)
        assert buf.size == 256
        assert buf.device_address >= 0
        assert not buf.freed
        assert not buf.mapped

    def test_allocation_with_usage(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            128,
            MemoryType.DEVICE_LOCAL,
            usage=BufferUsage.STORAGE | BufferUsage.TRANSFER_DST,
        )
        assert BufferUsage.STORAGE in buf.usage
        assert BufferUsage.TRANSFER_DST in buf.usage

    def test_host_visible_allocation(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64,
            MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT,
        )
        assert MemoryType.HOST_VISIBLE in buf.memory_type

    def test_unique_buffer_ids(self) -> None:
        device = make_device()
        mm = device.memory_manager
        b1 = mm.allocate(64, MemoryType.DEVICE_LOCAL)
        b2 = mm.allocate(64, MemoryType.DEVICE_LOCAL)
        assert b1.buffer_id != b2.buffer_id

    def test_stats_tracked(self) -> None:
        device = make_device()
        mm = device.memory_manager
        mm.allocate(256, MemoryType.DEVICE_LOCAL)
        mm.allocate(128, MemoryType.DEVICE_LOCAL)
        assert device.stats.total_allocations == 2
        assert device.stats.total_allocated_bytes == 384
        assert device.stats.peak_allocated_bytes == 384

    def test_invalid_size(self) -> None:
        device = make_device()
        mm = device.memory_manager
        with pytest.raises(ValueError, match="positive"):
            mm.allocate(0, MemoryType.DEVICE_LOCAL)

    def test_negative_size(self) -> None:
        device = make_device()
        mm = device.memory_manager
        with pytest.raises(ValueError, match="positive"):
            mm.allocate(-100, MemoryType.DEVICE_LOCAL)


class TestFree:
    def test_basic_free(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(256, MemoryType.DEVICE_LOCAL)
        mm.free(buf)
        assert buf.freed
        assert device.stats.total_frees == 1

    def test_double_free(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(256, MemoryType.DEVICE_LOCAL)
        mm.free(buf)
        with pytest.raises(ValueError, match="already freed"):
            mm.free(buf)

    def test_free_while_mapped(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mm.map(buf)
        with pytest.raises(ValueError, match="still mapped"):
            mm.free(buf)

    def test_current_bytes_after_free(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(256, MemoryType.DEVICE_LOCAL)
        assert mm.current_allocated_bytes == 256
        mm.free(buf)
        assert mm.current_allocated_bytes == 0

    def test_peak_bytes_after_free(self) -> None:
        device = make_device()
        mm = device.memory_manager
        b1 = mm.allocate(256, MemoryType.DEVICE_LOCAL)
        b2 = mm.allocate(128, MemoryType.DEVICE_LOCAL)
        mm.free(b1)
        assert device.stats.peak_allocated_bytes == 384
        assert mm.current_allocated_bytes == 128


class TestMap:
    def test_map_host_visible(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mapped = mm.map(buf)
        assert isinstance(mapped, MappedMemory)
        assert buf.mapped
        assert device.stats.total_maps == 1

    def test_map_device_local_fails(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(64, MemoryType.DEVICE_LOCAL)
        with pytest.raises(ValueError, match="HOST_VISIBLE"):
            mm.map(buf)

    def test_map_freed_fails(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mm.free(buf)
        with pytest.raises(ValueError, match="freed"):
            mm.map(buf)

    def test_double_map_fails(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mm.map(buf)
        with pytest.raises(ValueError, match="already mapped"):
            mm.map(buf)

    def test_unmap(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mm.map(buf)
        mm.unmap(buf)
        assert not buf.mapped

    def test_unmap_not_mapped_fails(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        with pytest.raises(ValueError, match="not mapped"):
            mm.unmap(buf)

    def test_unified_memory_map(self) -> None:
        """Apple unified: DEVICE_LOCAL + HOST_VISIBLE works."""
        device = make_device("apple")
        mm = device.memory_manager
        buf = mm.allocate(
            64,
            MemoryType.DEVICE_LOCAL | MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT,
        )
        mapped = mm.map(buf)
        assert mapped is not None
        mm.unmap(buf)


class TestMappedMemory:
    def test_read_write(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mapped = mm.map(buf)
        mapped.write(0, b"\x42" * 16)
        data = mapped.read(0, 16)
        assert data == b"\x42" * 16

    def test_write_at_offset(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mapped = mm.map(buf)
        mapped.write(32, b"\xAA" * 8)
        data = mapped.read(32, 8)
        assert data == b"\xAA" * 8

    def test_read_out_of_bounds(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            16, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mapped = mm.map(buf)
        with pytest.raises(ValueError, match="out of bounds"):
            mapped.read(0, 32)

    def test_write_out_of_bounds(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            16, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mapped = mm.map(buf)
        with pytest.raises(ValueError, match="out of bounds"):
            mapped.write(0, b"\x00" * 32)

    def test_dirty_flag(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mapped = mm.map(buf)
        assert not mapped.dirty
        mapped.write(0, b"\x01")
        assert mapped.dirty

    def test_get_data(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            8, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mapped = mm.map(buf)
        mapped.write(0, b"\x01\x02\x03\x04")
        data = mapped.get_data()
        assert len(data) == 8
        assert data[:4] == b"\x01\x02\x03\x04"

    def test_size_property(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            128, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mapped = mm.map(buf)
        assert mapped.size == 128


class TestFlushInvalidate:
    def test_flush(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mapped = mm.map(buf)
        mapped.write(0, b"\xFF" * 64)
        mm.flush(buf)  # Should not raise
        mm.unmap(buf)

    def test_invalidate(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mm.invalidate(buf)  # Should not raise

    def test_flush_freed_fails(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mm.free(buf)
        with pytest.raises(ValueError, match="freed"):
            mm.flush(buf)

    def test_invalidate_freed_fails(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(
            64, MemoryType.HOST_VISIBLE | MemoryType.HOST_COHERENT
        )
        mm.free(buf)
        with pytest.raises(ValueError, match="freed"):
            mm.invalidate(buf)


class TestBufferCount:
    def test_allocated_buffer_count(self) -> None:
        device = make_device()
        mm = device.memory_manager
        assert mm.allocated_buffer_count == 0
        b1 = mm.allocate(64, MemoryType.DEVICE_LOCAL)
        assert mm.allocated_buffer_count == 1
        b2 = mm.allocate(64, MemoryType.DEVICE_LOCAL)
        assert mm.allocated_buffer_count == 2
        mm.free(b1)
        assert mm.allocated_buffer_count == 1

    def test_get_buffer(self) -> None:
        device = make_device()
        mm = device.memory_manager
        buf = mm.allocate(64, MemoryType.DEVICE_LOCAL)
        retrieved = mm.get_buffer(buf.buffer_id)
        assert retrieved is buf

    def test_get_buffer_not_found(self) -> None:
        device = make_device()
        mm = device.memory_manager
        with pytest.raises(ValueError, match="not found"):
            mm.get_buffer(9999)

File: code/packages/ruby/compute_runtime/test/test_memory.rb
# frozen_string_literal: true

require_relative "test_helper"

# Helper to create a logical device for testing.
module MemoryTestHelper
  include CodingAdventures

  def make_device(vendor = "nvidia")
    instance = ComputeRuntime::RuntimeInstance.new
    physical = instance.enumerate_physical_devices.find { |d| d.vendor == vendor }
    instance.create_logical_device(physical)
  end
end

class TestAllocate < Minitest::Test
  include MemoryTestHelper

  def test_basic_allocation
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(256, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    assert_equal 256, buf.size
    assert buf.device_address >= 0
    refute buf.freed
    refute buf.mapped
  end

  def test_allocation_with_usage
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      128,
      ComputeRuntime::MemoryType::DEVICE_LOCAL,
      usage: ComputeRuntime::BufferUsage::STORAGE | ComputeRuntime::BufferUsage::TRANSFER_DST
    )
    assert (buf.usage & ComputeRuntime::BufferUsage::STORAGE) != 0
    assert (buf.usage & ComputeRuntime::BufferUsage::TRANSFER_DST) != 0
  end

  def test_host_visible_allocation
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64,
      ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    assert (buf.memory_type & ComputeRuntime::MemoryType::HOST_VISIBLE) != 0
  end

  def test_unique_buffer_ids
    device = make_device
    mm = device.memory_manager
    b1 = mm.allocate(64, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    b2 = mm.allocate(64, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    refute_equal b1.buffer_id, b2.buffer_id
  end

  def test_stats_tracked
    device = make_device
    mm = device.memory_manager
    mm.allocate(256, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    mm.allocate(128, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    assert_equal 2, device.stats.total_allocations
    assert_equal 384, device.stats.total_allocated_bytes
    assert_equal 384, device.stats.peak_allocated_bytes
  end

  def test_invalid_size
    device = make_device
    mm = device.memory_manager
    assert_raises(ArgumentError) do
      mm.allocate(0, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    end
  end

  def test_negative_size
    device = make_device
    mm = device.memory_manager
    assert_raises(ArgumentError) do
      mm.allocate(-100, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    end
  end
end

class TestFreeMemory < Minitest::Test
  include MemoryTestHelper

  def test_basic_free
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(256, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    mm.free(buf)
    assert buf.freed
    assert_equal 1, device.stats.total_frees
  end

  def test_double_free
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(256, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    mm.free(buf)
    assert_raises(ArgumentError) { mm.free(buf) }
  end

  def test_free_while_mapped
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mm.map(buf)
    assert_raises(ArgumentError) { mm.free(buf) }
  end

  def test_current_bytes_after_free
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(256, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    assert_equal 256, mm.current_allocated_bytes
    mm.free(buf)
    assert_equal 0, mm.current_allocated_bytes
  end

  def test_peak_bytes_after_free
    device = make_device
    mm = device.memory_manager
    b1 = mm.allocate(256, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    b2 = mm.allocate(128, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    mm.free(b1)
    assert_equal 384, device.stats.peak_allocated_bytes
    assert_equal 128, mm.current_allocated_bytes
  end
end

class TestMap < Minitest::Test
  include MemoryTestHelper

  def test_map_host_visible
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mapped = mm.map(buf)
    assert_kind_of ComputeRuntime::MappedMemory, mapped
    assert buf.mapped
    assert_equal 1, device.stats.total_maps
  end

  def test_map_device_local_fails
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(64, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    assert_raises(ArgumentError) { mm.map(buf) }
  end

  def test_map_freed_fails
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mm.free(buf)
    assert_raises(ArgumentError) { mm.map(buf) }
  end

  def test_double_map_fails
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mm.map(buf)
    assert_raises(ArgumentError) { mm.map(buf) }
  end

  def test_unmap
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mm.map(buf)
    mm.unmap(buf)
    refute buf.mapped
  end

  def test_unmap_not_mapped_fails
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    assert_raises(ArgumentError) { mm.unmap(buf) }
  end

  def test_unified_memory_map
    device = make_device("apple")
    mm = device.memory_manager
    buf = mm.allocate(
      64,
      ComputeRuntime::MemoryType::DEVICE_LOCAL |
        ComputeRuntime::MemoryType::HOST_VISIBLE |
        ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mapped = mm.map(buf)
    refute_nil mapped
    mm.unmap(buf)
  end
end

class TestMappedMemory < Minitest::Test
  include MemoryTestHelper

  def test_read_write
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mapped = mm.map(buf)
    mapped.write(0, "\x42".b * 16)
    data = mapped.read(0, 16)
    assert_equal "\x42".b * 16, data
  end

  def test_write_at_offset
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mapped = mm.map(buf)
    mapped.write(32, "\xAA".b * 8)
    data = mapped.read(32, 8)
    assert_equal "\xAA".b * 8, data
  end

  def test_read_out_of_bounds
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      16, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mapped = mm.map(buf)
    assert_raises(ArgumentError) { mapped.read(0, 32) }
  end

  def test_write_out_of_bounds
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      16, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mapped = mm.map(buf)
    assert_raises(ArgumentError) { mapped.write(0, "\x00".b * 32) }
  end

  def test_dirty_flag
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mapped = mm.map(buf)
    refute mapped.dirty
    mapped.write(0, "\x01".b)
    assert mapped.dirty
  end

  def test_get_data
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      8, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mapped = mm.map(buf)
    mapped.write(0, "\x01\x02\x03\x04".b)
    data = mapped.get_data
    assert_equal 8, data.bytesize
    assert_equal "\x01\x02\x03\x04".b, data.byteslice(0, 4)
  end

  def test_size_property
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      128, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mapped = mm.map(buf)
    assert_equal 128, mapped.size
  end
end

class TestFlushInvalidate < Minitest::Test
  include MemoryTestHelper

  def test_flush
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mapped = mm.map(buf)
    mapped.write(0, "\xFF".b * 64)
    mm.flush(buf) # Should not raise
    mm.unmap(buf)
  end

  def test_invalidate
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mm.invalidate(buf) # Should not raise
  end

  def test_flush_freed_fails
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mm.free(buf)
    assert_raises(ArgumentError) { mm.flush(buf) }
  end

  def test_invalidate_freed_fails
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(
      64, ComputeRuntime::MemoryType::HOST_VISIBLE | ComputeRuntime::MemoryType::HOST_COHERENT
    )
    mm.free(buf)
    assert_raises(ArgumentError) { mm.invalidate(buf) }
  end
end

class TestBufferCount < Minitest::Test
  include MemoryTestHelper

  def test_allocated_buffer_count
    device = make_device
    mm = device.memory_manager
    assert_equal 0, mm.allocated_buffer_count
    b1 = mm.allocate(64, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    assert_equal 1, mm.allocated_buffer_count
    b2 = mm.allocate(64, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    assert_equal 2, mm.allocated_buffer_count
    mm.free(b1)
    assert_equal 1, mm.allocated_buffer_count
  end

  def test_get_buffer
    device = make_device
    mm = device.memory_manager
    buf = mm.allocate(64, ComputeRuntime::MemoryType::DEVICE_LOCAL)
    retrieved = mm.get_buffer(buf.buffer_id)
    assert_same buf, retrieved
  end

  def test_get_buffer_not_found
    device = make_device
    mm = device.memory_manager
    assert_raises(ArgumentError) { mm.get_buffer(9999) }
  end
end
